// testcafe-hammerhead/lib/request-pipeline/cache.js
"use strict";

exports.__esModule = true;
exports.shouldCache = shouldCache;
exports.create = create;
exports.add = add;
exports.getResponse = getResponse;
exports.MAX_SIZE_FOR_NON_PROCESSED_RESOURCE = void 0;

var _lruCache = _interopRequireDefault(require("lru-cache"));

var _httpCacheSemantics = _interopRequireDefault(require("http-cache-semantics"));

var _incomingMessageLike = _interopRequireDefault(require("./incoming-message-like"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const requestsCache = new _lruCache.default({
  max: 50 * 1024 * 1024,
  // Max cache size is 50 MBytes.
  length: responseCacheEntry => {
    // NOTE: Length is resource content size.
    // 1 character is 1 bite.
    return responseCacheEntry.res.getBody().length;
  }
});

function getCacheKey(requestOptions) {
  // NOTE: We don't use pair <method:url> as a cache key since we cache only GET requests
  return requestOptions.url;
} // NOTE: export for testing purposes


function shouldCache(ctx) {
  return ctx.serverInfo.cacheRequests && !ctx.isFileProtocol && ctx.reqOpts.method === 'GET' && (ctx.contentInfo.isCSS || ctx.contentInfo.isScript || !ctx.contentInfo.requireProcessing);
}

function create(reqOptions, res) {
  const cachePolicy = new _httpCacheSemantics.default(reqOptions, res);
  if (!cachePolicy.storable()) return void 0;
  return {
    key: getCacheKey(reqOptions),
    value: {
      cachePolicy,
      res: _incomingMessageLike.default.createFrom(res),
      hitCount: 0
    }
  };
}

function add(entry) {
  const {
    key,
    value
  } = entry;
  requestsCache.set(key, value, value.cachePolicy.timeToLive());
}

function getResponse(reqOptions) {
  const key = getCacheKey(reqOptions);
  const cachedResponse = requestsCache.get(key);
  if (!cachedResponse) return void 0;
  const {
    cachePolicy,
    res
  } = cachedResponse;
  if (!cachePolicy.satisfiesWithoutRevalidation(reqOptions)) return void 0;
  res.headers = cachePolicy.responseHeaders();
  cachedResponse.hitCount++;
  return {
    res,
    hitCount: cachedResponse.hitCount
  };
} // NOTE: Maximum size for the non processed resource is 5 Mb


const MAX_SIZE_FOR_NON_PROCESSED_RESOURCE = 5 * 1024 * 1024;
exports.MAX_SIZE_FOR_NON_PROCESSED_RESOURCE = MAX_SIZE_FOR_NON_PROCESSED_RESOURCE;
